### PR TITLE
OPS-4735 added optional Cloud Front functions

### PR DESCRIPTION
**Pull request by vselcuk**
# Optional Cloud Front functions

## Description
This PR adds optional Cloud Front functions creation and association with the created Cloud Front distribution so that we would be able to associate some logic to the CF request and response.

## Testing Instructions
The module is used to create a cdn to serve the product feeds generated by Productsup and our custom lambda function in the following lite-infra branch. The optional function creation would enable us to add CF functions to the distribution. Please see the usage here and it has been deployed to dev.

https://github.com/Flaconi/lite-infra/blob/PIMF-1502/terragrunt/envs/aws/dev/eu-central-1/applications/ct-product-feed/feed-cdn/terragrunt.hcl

## How to roll out
After PR approval the release tag is created and the relevant tag is updated in the branch PIMF-1502 of lite-infra repository.

## Demo
The CF function enables basic http auth for the following cdn url:
(https://product-feed.dev.flaconi-platform.net/export/pl_raw_data.txt)

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -115,3 +119,13 @@ resource "aws_route53_record" "this" {
     evaluate_target_health = false
   }
 }
+
+resource "aws_cloudfront_function" "functions" {
+  count = length(var.cf_functions)
```
Comment (snovikov) [inline main.tf:124]: Better to use `for_each` instead of count.

Comment (vselcuk) [reply main.tf:124]: The variable is a list and I think for_each does not work for the list type (at least for me), but I can change it like the following:
```
resource "aws_cloudfront_function" "functions" {
  for_each = {for func in var.cf_functions: func.type => func}

  name    = each.value.name
  runtime = "cloudfront-js-1.0"
  comment = each.value.comment
  publish = true
  code    = each.value.code
}
```

Comment (snovikov) [reply main.tf:124]: yes, I meant that. but keys should be unique.
```
for_each = {for func in var.cf_functions: func.name => func}
```

Comment (snovikov) [reply main.tf:124]: from the discussion it became clear that `func.type` should be an unique field.

Comment (vselcuk) [reply main.tf]: I have changed the data type of the input variable to map and have updated the relevant places

**Review by snovikov — APPROVED** [review]

**Pull request merged**